components/usage/pkg/controller/reconciler.go

```diff
@@ -38,16 +47,100 @@ func (u *UsageReconciler) Reconcile() error {
 	startOfCurrentMonth := time.Date(now.Year(), now.Month(), 1, 0, 0, 0, 0, time.UTC)
 	startOfNextMonth := startOfCurrentMonth.AddDate(0, 1, 0)
 
-	return u.reconcile(ctx, startOfCurrentMonth, startOfNextMonth)
+	status, err := u.ReconcileTimeRange(ctx, startOfCurrentMonth, startOfNextMonth)
+	if err != nil {
+		return err
+	}
+	log.WithField("usage_reconcile_status", status).Info("Reconcile completed.")
+	return nil
 }
 
-func (u *UsageReconciler) reconcile(ctx context.Context, from, to time.Time) error {
+func (u *UsageReconciler) ReconcileTimeRange(ctx context.Context, from, to time.Time) (*UsageReconcileStatus, error) {
 	log.Infof("Gathering usage data from %s to %s", from, to)
-	instances, err := db.ListWorkspaceInstancesInRange(ctx, u.conn, from, to)
+	status := &UsageReconcileStatus{
+		StartTime: from,
+		EndTime:   to,
+	}
+	instances, invalidInstances, err := u.loadWorkspaceInstances(ctx, from, to)
 	if err != nil {
-		return fmt.Errorf("failed to list instances: %w", err)
+		return nil, fmt.Errorf("failed to load workspace instances: %w", err)
 	}
+	status.WorkspaceInstances = len(instances)
+	status.InvalidWorkspaceInstances = len(invalidInstances)
 
+	if len(invalidInstances) > 0 {
+		log.WithField("invalid_workspace_instances", invalidInstances).Errorf("Detected %d invalid instances. These will be skipped in the current run.", len(invalidInstances))
+	}
+
+	log.WithField("workspace_instances", instances).Debug("Successfully loaded workspace instances.")
+	return status, nil
+}
+
+func (u *UsageReconciler) loadWorkspaceInstances(ctx context.Context, from, to time.Time) ([]db.WorkspaceInstance, []invalidWorkspaceInstance, error) {
+	log.Infof("Gathering usage data from %s to %s", from, to)
+	instances, err := db.ListWorkspaceInstancesInRange(ctx, u.conn, from, to)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to list instances from db: %w", err)
+	}
 	log.Infof("Identified %d instances between %s and %s", len(instances), from, to)
-	return nil
+
+	valid, invalid := validateInstances(instances)
+	trimmed := trimStartStopTime(valid, from, to)
+	return trimmed, invalid, nil
+}
+
+type invalidWorkspaceInstance struct {
+	reason              string
+	workspaceInstanceID uuid.UUID
+}
+
+func validateInstances(instances []db.WorkspaceInstance) (valid []db.WorkspaceInstance, invalid []invalidWorkspaceInstance) {
+	for _, i := range instances {
+		// i is a pointer to the current element, we need to assign it to ensure we're copying the value, not the current pointer.
+		instance := i
+
+		// Each instance must have a start time, without it, we do not have a baseline for usage computation.
+		if !instance.CreationTime.IsSet() {
+			invalid = append(invalid, invalidWorkspaceInstance{
+				reason:              "missing creation time",
```

Comment:
Need constants for these literals?

Reply:
These are only used in a single place and intended for human consumption so they can fix it up. I think we're fine with inlined.